lib/statsd/instrument/batched_udp_sink.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+module StatsD
+  module Instrument
+    # @note This class is part of the new Client implementation that is intended
+    #   to become the new default in the next major release of this library.
+    class BatchedUDPSink
+      DEFAULT_FLUSH_INTERVAL = 1.0
+      MAX_PACKET_SIZE = 508
+
+      def self.for_addr(addr, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        host, port_as_string = addr.split(":", 2)
+        new(host, Integer(port_as_string), flush_interval: flush_interval)
+      end
+
+      attr_reader :host, :port
+
+      def initialize(host, port, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        @host = host
+        @port = port
+        @mutex = Mutex.new
+        @socket = nil
+        @flush_interval = flush_interval
+
+        require "concurrent/array"
+        @buffer = Concurrent::Array.new
+        @dispatcher_thread = nil
+        spawn_dispatcher
+      end
+
+      def sample?(sample_rate)
+        sample_rate == 1 || rand < sample_rate
+      end
+
+      def <<(datagram)
+        @buffer << datagram
+        spawn_dispatcher unless @dispatcher_thread&.alive?
+        self
+      end
+
+      def shutdown
+        @dispatcher_thread&.kill
+      end
+
+      private
+
+      def spawn_dispatcher
+        @dispatcher_thread = Thread.new { dispatch }
+      end
+
+      NEWLINE = "\n".b.freeze
+      def flush
+        return if @buffer.empty?
+
+        datagrams = @buffer.shift(@buffer.size)
+
+        until datagrams.empty?
+          packet = String.new(datagrams.pop, encoding: Encoding::BINARY, capacity: MAX_PACKET_SIZE)
```

Comment:
Ideally, we would re-use the same string across packets, however, I don't see a ruby method for just setting the length to 0 (i.e. clearing the contents) without freeing the capacity.

Reply:
Yes, I went though the same idea. Unfortunately as you point it `String#clear` frees the capacity entirely.

I haven't tested `StringIO`, but from experience it tend to actually be slower than Strings.